lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2561,20 +2560,31 @@ namespace {
         }
 
         swift::NominalTypeDecl *paramDecl = paramType->getAnyNominal();
+
+        // The return type should be void (for release functions), or void
+        // or the parameter type (for retain functions).
+        auto resultInterfaceType = operationFn->getResultInterfaceType();
+        if (!resultInterfaceType->isVoid()) {
+          if (operationKind == CustomRefCountingOperationKind::release ||
+              resultInterfaceType->getAnyNominal() != paramDecl)
```

Comment:
This check isn't correct: it's comparing the declaration result's nominal type (if there is one) against the parameter declaration itself (which has, e.g., the name of the parameter in it). Instead, you need to compare the result type against the parameter type. I also expect that we need to look through an optional type on the result type, in case the retain function handles null pointers in some manner. Specifically, I think it should be:

```suggestion
              !resultInterfaceType->lookThroughSingleOptionalType()->isEqual(paramType))
```

Reply:
I'm confused about why this PR worked without this change suggested above. Running tests in CI without the change because I would expect it to fail.

Reply:
Fix applied (shouldn't disrupt running builds to check previous commit).